### DISCUSSION TITLE
is the keycloak-servlet filter as well being deprecated

### DISCUSSION THREAD

**Discussion by kkcmadhu** [keycloak/keycloak#11790] [Adapter deprecation]
I use this  keycloak adapter https://github.com/keycloak/keycloak/tree/main/adapters/oidc/servlet-filter

It was not very clear where this one is getting deprecated or not. Can some one please clarify.
In case this is being deprecated, what is the alternative one should consider?

**Comment by kkcmadhu** [comment]
Ah.. had a second look here https://www.keycloak.org/2022/02/adapter-deprecation.html , the servlet filter is  not being  deprecated.

**Comment by stianst** [comment]
The SAML servlet filter will remain, but the OIDC servlet filter is planned to be removed.

**Comment by kkcmadhu** [comment]
What's the alternative for oidc srvlet filter ( for non spring application)

Sent from Yahoo Mail on Android 
 
  On Wed, 4 May 2022 at 4:33 pm, Stian ***@***.***> wrote:   


The SAML servlet filter will remain, but the OIDC servlet filter is planned to be removed.

**Reply by kkcmadhu** [reply]
quick look at various alternatives suggests google oauth 2 client library for servlets (https://developers.google.com/api-client-library/java/google-oauth-java-client/setup)

there are few others form other oauth providers like https://auth0.com/docs/quickstart/webapp/java/01-login or  oktas (https://github.com/okta/okta-auth-java) etc..
But again these are  client library provided by  differnet oAuth providers, i **did  not find anything truly generic  (like what you have for spring boot etc).**

@stianst  I think it might be worth while maintaining the kyecloak servlet libraries which is fully compatible with keycloak.. Let me know your opinion.

**Comment by ssilvert** [comment]
@stianst If we really can't find a good generic replacement, then I would tend to agree with @kkcmadhu.  We do have some good servlet expertise on staff (me) if we need to maintain this a little longer.

**Reply by kkcmadhu** [reply]
thanks much @ssilvert.

**Reply by stianst** [reply]
We are not able to maintain the servlet filter beyond what is laid out in https://www.keycloak.org/2022/03/releases. The cost of maintaining this is very big compared to the limited number of users.

An option is of course to extract the servlet filter (and any other adapters) into a separate community maintained repository if anyone wants to maintain this beyond March/April 2023.

**Reply by kiswanij** [reply]
Thanks @ssilvert , maintaining it for a longer time and maybe updating to the jakarta namespace (package) instead of javax would be great for us.

**Reply by ShubhSingh** [reply]
@stianst even I would very much like to have OIDC servlet filter maintained for long term.

**Reply by gjain-sag** [reply]
Hi @stianst 

Was going through this discussion and understood about keycloak's plan to discontinue suport for oidc-servlet client libraries.
Currently in my org several applications are still on tomcat servlet architecture and we are heavely dependent on keycloaks oidc-servlet client adapter. 
I echo with @kkcmadhu there are not much open source alternative available which support all the use cases as of now. So we request to contiue support for oidc-servlet library a longer till a good open-source alternative is available.

**Reply by kkcmadhu** [reply]
> g it for a longer time and maybe updating to the jakarta namespace (package) instead of javax would be great for us.

i see [quarkus3 branch already doing it 
](https://github.com/keycloak/keycloak/commit/c4dc85ec53c6ce26ee3b5a1f538955b6643bd59d#diff-21878817a5ea36c762a1605356de3f7f5949684275ff56e4c72aa612516f34ae)
including the [oidc-servlet-library modules ](https://github.com/keycloak/keycloak/commit/c4dc85ec53c6ce26ee3b5a1f538955b6643bd59d#diff-03cd8807421128273e1d857a1ea1bf838664df46fdbc6a006acdde010020aa26)

**Comment by kkcmadhu** [comment]
I did some more digging around the servlet filter deprecation , I think  this will impact the following use cases

[always refresh token](https://www.keycloak.org/docs/latest/securing_apps/index.html#_refresh_token_each_req)

[Clustering/ node registration](https://www.keycloak.org/docs/latest/securing_apps/index.html#_applicationclustering)

Many of the K_ events /back channel logout , logout propogation etc..

and even [multitenancy]( https://www.keycloak.org/docs/latest/securing_apps/index.html#_multi_tenancy)

Also there are host of client adapter settings which are not directly portable/compatible
E.g.
ssl-required ,confidential-port,use-resource-role-mappings,autodetect-bearer-only,expose-token,always-refresh-token,register-node-at-startup,verify-token-audience etc..



Many applications still uses tomcat/servlet etc..
While at one level, It  may make sense to deprecate  tomcat, jetty,wildfy adapters,spring boot .. but something as fundamental as java servlet filter should definately be supported in my opnion.



Unfortunately, There is no good generic replacment library avaialble for oidc-servlet-filter in my opinion. :(

**Reply by movertaker** [reply]
You are right. We have the same problem. At the moment, we plan to migrate our applications from tomcat/tomee to wildfly for this reason.

**Comment by stianst** [comment]
Thanks for all the inputs. Based on feedback we have decided to revise our plans around adapter deprecation and removal. For the updated plans please see https://www.keycloak.org/2023/03/adapter-deprecation-update.html.

**Reply by kkcmadhu** [reply]
Thanks @stianst  much appriciated .It was really kind of you to revaulate the plan and to offer support for these adapters a while more!

**Comment by col-panic** [comment]
Maybe this is a solution, at least if you only want to protect an endpoint https://github.com/curityio/oauth-filter-for-java
